src/frontend/http.rs

```diff
@@ -64,138 +75,279 @@ struct QueryBody {
+pub async fn upload(
+    schema_name: String,
+    table_name: String,
+    form: FormData,
+    context: Arc<dyn SeafowlContext>,
+) -> Response {
+    let parts: Vec<Part> = form.try_collect().await.unwrap();
+
+    let mut has_header = true;
+    let mut csv_schema: Option<Schema> = None;
+    let mut filename = String::new();
+    for p in parts {
+        if p.name() == "has_header"
+            || p.name() == "schema" && p.content_type() == Some("application/json")
+        {
+            let header_part = p.name() == "has_header";
+
+            let value_bytes = p
+                .stream()
+                .try_fold(Vec::new(), |mut vec, data| {
+                    vec.put(data);
+                    async move { Ok(vec) }
+                })
+                .await
+                .map_err(|e| {
+                    eprintln!("Error reading part's data: {}", e);
+                    warp::reject::reject()
+                })
+                .unwrap();
+
+            if header_part {
+                has_header = String::from_utf8(value_bytes).unwrap().starts_with("true");
+                debug!("Form part has_header is: {}", has_header);
+            } else {
+                let schema_json: Value =
+                    serde_json::from_slice(value_bytes.as_slice()).unwrap();
+                csv_schema = Some(Schema::from(&schema_json).unwrap());
+            }
+        } else if p.name() == "data" || p.name() == "file" {
+            filename = p.filename().unwrap().to_string();
+
+            // Load the file content from the request
+            // TODO: we're actually buffering the entire file into memory here which is sub-optimal,
+            // since we could be writing the contents of the stream out into a temp file on the disk
+            // to have a smaller memory footprint. However, for this to be supported warp first
+            // needs to support streaming here: https://github.com/seanmonstar/warp/issues/323
+            let value = p
+                .stream()
+                .try_fold(Vec::new(), |mut vec, data| {
+                    vec.put(data);
+                    async move { Ok(vec) }
+                })
+                .await
+                .map_err(|e| {
+                    eprintln!("Error reading part's data: {}", e);
+                    warp::reject::reject()
+                })
+                .unwrap();
+
+            // Parse the schema and load the file contents into a vector of record batches
+            let (schema, partition) = match filename.split('.').last().unwrap() {
+                "csv" => {
+                    // If the schema part wasn't specified we'll need to infer it
+                    let builder = match csv_schema.clone() {
+                        Some(schema) => ReaderBuilder::new()
+                            .with_schema(Arc::new(schema.clone()))
+                            .has_header(has_header), // .with_escape(b'\\'); // default is None, change to \
```

Comment:
If in the future we want to let the user to be able to override some other CSV options like the separator, will we want to continue using the formdata for it or should we use some other way? I had this other idea which might be more appropriate in that case: we let the user pass in a `CREATE EXTERNAL TABLE` query that references the freshly uploaded dataset, e.g.

```sql
CREATE EXTERNAL TABLE my_csv_table ( some: INTEGER, schema: INTEGER)
STORED AS CSV
WITH HEADER ROW, DELIMITER '\t'
LOCATION "@data"
```

similar to Snowflake stages (https://docs.snowflake.com/en/user-guide/querying-stage.html) but it needs some more thinking/API design.

Reply:
Yup, I would say that looks like a suitable use case for formdata as well (as in pass the query in another part of the multipart request with e.g. `-F "query=...").